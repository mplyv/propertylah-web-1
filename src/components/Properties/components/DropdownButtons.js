import styles from '../property.module.css';

const DropdownButtons = (props)=>{
    return(
        props.type === 'num'? 
        <div className={styles.horizontalContainerPopup}>
            <div data-testid={props.options[0]} id={props.options[0]} className={props.variable.includes('All')? styles.selected : styles.filterSearchBarButton} onClick={props.handleChange}>All</div>
            {props.options.map((option)=>{
                return(
                    option.includes('All')? null : option.includes('+')? 
                    <div data-testid={option} key={option} id={option} className={option.includes(props.variable)? styles.selected:styles.filterSearchBarButton} onClick={props.handleChange}>{option.slice(0,2)}</div> :
                    <div data-testid={option} key={option} id={option} className={option.includes(props.variable)? styles.selected:styles.filterSearchBarButton} onClick={props.handleChange}>{option.slice(0,1)}</div>
                )
            }
            )}
        </div>
        :
        <div className={styles.horizontalContainerPopup}>
            <div data-testid={props.options[0]} id={props.options[0]} className={props.variable.includes('All')? styles.selected : styles.filterSearchBarButton} onClick={props.handleChange}>All</div>
            {props.options.map((option)=>{
                return(
                    option.includes('All')? null:<div data-testid={option} key={option} id={option} className={option.includes(props.variable)? styles.selected:styles.filterSearchBarButton} onClick={props.handleChange}>{option}</div>
                )
            }
            )}
        </div>
    )
}
    
export default DropdownButtons