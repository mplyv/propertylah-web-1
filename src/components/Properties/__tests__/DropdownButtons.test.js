import {render, screen, cleanup} from '@testing-library/react';
import renderer from 'react-test-renderer'
import DropdownButtons from '../components/DropdownButtons'
import userEvent from '@testing-library/user-event'


//* cleanup after each test
afterEach(() => {
  cleanup();
});

let propertyTypeOptions = ['All Property','HDB','Apartment','Condominium'];
let propertyType = 'All'
let bathOptions = ['All Qty Bathrooms','1bath','2bath','3bath','4bath','5+bath'];
let noOfBathrooms = 'All'
let bedroomOptions = ['All Qty Bedrooms','1bed','2bed','3bed','4bed','5+bed'];
let noOfBedrooms = 'All'
let tenureOptions = ['All Tenure','Freehold','Leasehold'];
let tenure = 'All'

test('Should render a div for each option for type=(text/num)',()=>{
    const {rerender} = render(<DropdownButtons handleChange={jest.fn()} options={propertyTypeOptions} type='text' variable={propertyType}/>);
    expect(screen.getByTestId('All Property')).toBeInTheDocument();
    expect(screen.getByTestId('HDB')).toBeInTheDocument();
    expect(screen.getByTestId('Apartment')).toBeInTheDocument();
    expect(screen.getByTestId('Condominium')).toBeInTheDocument();

    rerender(<DropdownButtons handleChange={jest.fn()} options={bathOptions} type='num' variable={noOfBathrooms}/>);
    expect(screen.getByTestId('All Qty Bathrooms')).toBeInTheDocument();
    expect(screen.getByTestId('1bath')).toBeInTheDocument();
    expect(screen.getByTestId('2bath')).toBeInTheDocument();
    expect(screen.getByTestId('3bath')).toBeInTheDocument();
    expect(screen.getByTestId('4bath')).toBeInTheDocument();
    expect(screen.getByTestId('5+bath')).toBeInTheDocument();
});

test('Should render divs with innerHTML according to options',()=>{
    const {rerender} = render(<DropdownButtons handleChange={jest.fn()} options={propertyTypeOptions} type='text' variable={propertyType}/>);
    expect(screen.getByTestId('All Property').innerHTML).toBe('All');
    expect(screen.getByTestId('HDB').innerHTML).toBe('HDB');
    expect(screen.getByTestId('Apartment').innerHTML).toBe('Apartment');
    expect(screen.getByTestId('Condominium').innerHTML).toBe('Condominium');

    rerender(<DropdownButtons handleChange={jest.fn()} options={tenureOptions} type='text' variable={tenure}/>);
    expect(screen.getByTestId('All Tenure').innerHTML).toBe('All');
    expect(screen.getByTestId('Freehold').innerHTML).toBe('Freehold');
    expect(screen.getByTestId('Leasehold').innerHTML).toBe('Leasehold');
});

test('1bath to 4bath should render as just 1,2,3,4 // 5+bath should render as 5+' ,()=>{
    render(<DropdownButtons handleChange={jest.fn()} options={bathOptions} type='num' variable={noOfBathrooms}/>);
    expect(screen.getByTestId('1bath').innerHTML).toBe('1');
    expect(screen.getByTestId('2bath').innerHTML).toBe('2');
    expect(screen.getByTestId('3bath').innerHTML).toBe('3');
    expect(screen.getByTestId('4bath').innerHTML).toBe('4');
    expect(screen.getByTestId('5+bath').innerHTML).toBe('5+');
});

test('All Qty Bathrooms should render as All',()=>{
    render(<DropdownButtons handleChange={jest.fn()} options={bathOptions} type='num' variable={noOfBathrooms}/>);
    expect(screen.getByTestId('All Qty Bathrooms').innerHTML).toBe('All');
});

// className tests
test('className for All should be (selected) by default, other buttons should have className(filterSearchBarButton) by default',()=>{
    render(<DropdownButtons handleChange={jest.fn()} options={bathOptions} type='num' variable={noOfBathrooms}/>);
    expect(screen.getByTestId('All Qty Bathrooms').className).toBe('selected');
    expect(screen.getByTestId('1bath').className).toBe('filterSearchBarButton');
    expect(screen.getByTestId('2bath').className).toBe('filterSearchBarButton');
    expect(screen.getByTestId('3bath').className).toBe('filterSearchBarButton');
    expect(screen.getByTestId('4bath').className).toBe('filterSearchBarButton');
});

let handleNoOfBaths = jest.fn();
test('handleNoOfBaths function called when pressing buttons', async()=>{
    render(<DropdownButtons handleChange={handleNoOfBaths} options={bathOptions} type='num' variable={noOfBathrooms}/>);
    const user = userEvent.setup()
    await user.click(screen.getByTestId('1bath'));
    expect(handleNoOfBaths).toHaveBeenCalled();
});

test('className changes when pressing buttons', async()=>{
    const {rerender} = render(<DropdownButtons handleChange={handleNoOfBaths} options={bathOptions} type='num' variable={'1'}/>);
    expect(screen.getByTestId('1bath').className).toBe('selected');
    expect(screen.getByTestId('All Qty Bathrooms').className).toBe('filterSearchBarButton');

    rerender(<DropdownButtons handleChange={handleNoOfBaths} options={bathOptions} type='num' variable={'2'}/>)
    expect(screen.getByTestId('2bath').className).toBe('selected');
    expect(screen.getByTestId('All Qty Bathrooms').className).toBe('filterSearchBarButton');
    rerender(<DropdownButtons handleChange={handleNoOfBaths} options={bathOptions} type='num' variable={'3'}/>)
    expect(screen.getByTestId('3bath').className).toBe('selected');
    expect(screen.getByTestId('All Qty Bathrooms').className).toBe('filterSearchBarButton');
    rerender(<DropdownButtons handleChange={handleNoOfBaths} options={bathOptions} type='num' variable={'4'}/>)
    expect(screen.getByTestId('4bath').className).toBe('selected');
    expect(screen.getByTestId('All Qty Bathrooms').className).toBe('filterSearchBarButton');
    rerender(<DropdownButtons handleChange={handleNoOfBaths} options={bathOptions} type='num' variable={'5'}/>)
    expect(screen.getByTestId('5+bath').className).toBe('selected');
    expect(screen.getByTestId('All Qty Bathrooms').className).toBe('filterSearchBarButton');
});


//Snapshot tests
test('match bathroom snapshot',()=>{
    const tree = renderer.create(<DropdownButtons handleChange={jest.fn()} options={bathOptions} type='num' variable={noOfBathrooms}/>).toJSON();
    expect(tree).toMatchSnapshot();
})

test('match bedrooms snapshot',()=>{
    const tree = renderer.create(<DropdownButtons handleChange={jest.fn()} options={bedroomOptions} type='num' variable={noOfBedrooms}/>).toJSON();
    expect(tree).toMatchSnapshot();
})

test('match propertyType snapshot',()=>{
    const tree = renderer.create(<DropdownButtons handleChange={jest.fn()} options={propertyTypeOptions} type='text' variable={propertyType}/>).toJSON();
    expect(tree).toMatchSnapshot();
})

test('match tenure snapshot',()=>{
    const tree = renderer.create(<DropdownButtons handleChange={jest.fn()} options={tenureOptions} type='text' variable={tenure}/>).toJSON();
    expect(tree).toMatchSnapshot();
})