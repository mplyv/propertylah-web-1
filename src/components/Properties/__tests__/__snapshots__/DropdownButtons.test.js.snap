// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match bathroom snapshot 1`] = `
<div
  className="horizontalContainerPopup"
>
  <div
    className="selected"
    data-testid="All Qty Bathrooms"
    id="All Qty Bathrooms"
    onClick={[MockFunction]}
  >
    All
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="1bath"
    id="1bath"
    onClick={[MockFunction]}
  >
    1
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="2bath"
    id="2bath"
    onClick={[MockFunction]}
  >
    2
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="3bath"
    id="3bath"
    onClick={[MockFunction]}
  >
    3
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="4bath"
    id="4bath"
    onClick={[MockFunction]}
  >
    4
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="5+bath"
    id="5+bath"
    onClick={[MockFunction]}
  >
    5+
  </div>
</div>
`;

exports[`match bedrooms snapshot 1`] = `
<div
  className="horizontalContainerPopup"
>
  <div
    className="selected"
    data-testid="All Qty Bedrooms"
    id="All Qty Bedrooms"
    onClick={[MockFunction]}
  >
    All
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="1bed"
    id="1bed"
    onClick={[MockFunction]}
  >
    1
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="2bed"
    id="2bed"
    onClick={[MockFunction]}
  >
    2
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="3bed"
    id="3bed"
    onClick={[MockFunction]}
  >
    3
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="4bed"
    id="4bed"
    onClick={[MockFunction]}
  >
    4
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="5+bed"
    id="5+bed"
    onClick={[MockFunction]}
  >
    5+
  </div>
</div>
`;

exports[`match propertyType snapshot 1`] = `
<div
  className="horizontalContainerPopup"
>
  <div
    className="selected"
    data-testid="All Property"
    id="All Property"
    onClick={[MockFunction]}
  >
    All
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="HDB"
    id="HDB"
    onClick={[MockFunction]}
  >
    HDB
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="Apartment"
    id="Apartment"
    onClick={[MockFunction]}
  >
    Apartment
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="Condominium"
    id="Condominium"
    onClick={[MockFunction]}
  >
    Condominium
  </div>
</div>
`;

exports[`match tenure snapshot 1`] = `
<div
  className="horizontalContainerPopup"
>
  <div
    className="selected"
    data-testid="All Tenure"
    id="All Tenure"
    onClick={[MockFunction]}
  >
    All
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="Freehold"
    id="Freehold"
    onClick={[MockFunction]}
  >
    Freehold
  </div>
  <div
    className="filterSearchBarButton"
    data-testid="Leasehold"
    id="Leasehold"
    onClick={[MockFunction]}
  >
    Leasehold
  </div>
</div>
`;
